### PR TITLE
adds function type for Dictionary version of reduce (issue 4244)

### PR DESCRIPTION
**Pull request by dbeckwith**
Added new interface that basically combines ObjectIterator and MemoIterator.

### CODE REVIEW

**Comment by vvakame** [comment]
@davetayls could you review this PR?

File: underscore/underscore.d.ts [hunk 1]
```diff
@@ -53,6 +53,10 @@ declare module _ {
 		(prev: TResult, curr: T, index: number, list: T[]): TResult;
 	}
 
+	interface MemoObjectIterator<T, TResult> {
+		(prev: TResult, curr: T, key: string, list: any): TResult;
```
Comment (davetayls) [inline underscore/underscore.d.ts]: Should the list be a _.Dictionary<T>

**Comment by dbeckwith** [comment]
I was just copying the signature from MemoIterator by making it any

**Comment by dbeckwith** [comment]
I think you're right though, and above in ListIterator it should be _.List. I can change that, and I'll add some more tests.

**Comment by davetayls** [comment]
Looking good to me @vvakame

**Comment by vvakame** [comment]
@davetayls thanks!
@dbeckwith thanks mate!

**Comment by davetayls** [comment]
no worries

**Pull request merged**